verilog/tools/ls/symbol-table-handler.cc

```diff
@@ -0,0 +1,177 @@
+    absl::string_view corpus) {
+  currproject = std::make_unique<VerilogProject>(root, include_paths, corpus);
+  resetSymbolTable();
+}
+
+void SymbolTableHandler::resetSymbolTable() {
+  checkedfiles.clear();
+  symboltable = std::make_unique<SymbolTable>(currproject.get());
+}
+
+void SymbolTableHandler::buildSymbolTableFor(VerilogSourceFile &file) {
+  auto result = BuildSymbolTable(file, symboltable.get(), currproject.get());
+}
+
+void SymbolTableHandler::buildProjectSymbolTable() {
+  resetSymbolTable();
+  if (!currproject) {
+    return;
+  }
+  LOG(INFO) << "Parsing project files...";
+  for (const auto &file : *currproject) {
+    auto status = file.second->Parse();
+    if (!status.ok()) {
+      LOG(ERROR) << "Failed to parse file:  " << file.second->ReferencedPath();
+      return;
+    }
+    LOG(INFO) << "Successfully parsed:  " << file.second->ReferencedPath();
+    auto result =
+        BuildSymbolTable(*file.second, symboltable.get(), currproject.get());
+  }
+  LOG(INFO) << "Parsed project files";
+  LOG(INFO) << "Symbol table for the project";
+  symboltable->PrintSymbolDefinitions(std::cerr);
+}
+
+const SymbolTableNode *SymbolTableHandler::ScanSymbolTreeForDefinition(
+    const SymbolTableNode *context, absl::string_view symbol) {
+  if (!context) {
+    return nullptr;
+  }
+  if (context->Key() && *context->Key() == symbol) {
+    return context;
+  }
+  for (const auto &child : context->Children()) {
+    auto res = ScanSymbolTreeForDefinition(&child.second, symbol);
+    if (res) {
+      return res;
+    }
+  }
+  LOG(INFO) << "Symbol definitions for the project:";
+  symboltable->PrintSymbolDefinitions(std::cerr);
+  return nullptr;
+}
+
+std::vector<verible::lsp::Location> SymbolTableHandler::findDefinition(
+    const verible::lsp::DefinitionParams &params,
+    const verilog::BufferTrackerContainer &parsed_buffers) {
+  std::string filepath;
+  if (!LSPUriToPath(params.textDocument.uri, &filepath)) {
+    std::cerr << "Could not convert URI " << params.textDocument.uri
+              << " to filesystem path." << std::endl;
+    return {};
+  }
+  auto relativepath = currproject->GetRelativePathToSource(filepath);
+  // TODO add checking parsed buffers / raw buffers to get the newest state of
+  // files and fallback to reading files from filesystem only when necessary.
+  if (checkedfiles.find(relativepath) == checkedfiles.end()) {
+    // File hasn't been tracked yet in the symbol table, add it
+    auto openedfile = currproject->OpenTranslationUnit(relativepath);
+    // TODO check parse status
+    auto parsestatus = (*openedfile)->Parse();
+    if (!openedfile.ok()) {
+      LOG(WARNING) << "Could not open [" << filepath << "] in project ["
+                   << currproject->TranslationUnitRoot() << "]";
+      return {};
+    }
+    auto buildstatus =
+        BuildSymbolTable(**openedfile, symboltable.get(), currproject.get());
+  }
+  auto parsedbuffer =
```

Comment:
If it is not too long, I'd suggest to use the full type
```
const BufferTracker *parsedbuffer = ...
```

While `auto` is quick to write and often shorter, it makes understanding the code often harder as the reader of the code does not see the type.

My personal way of dealing with it is to typically use the fully qualified type, unless it is somewhat bulky and somewhat clear from the context (e.g. say a for iterators).